### PR TITLE
Add support for standard crontab format

### PR DESCRIPTION
**Pull request by kylesnowschwartz**
# Add support for standard crontab format

## Summary
- Add support for the standard unquoted crontab format that's used in real-world crontab files
- Implement a new `cron_from_line_crontab` function that recognizes both quoted and unquoted expressions
- Add comprehensive acceptance tests for the new functionality
- Include example configuration for users

## Motivation
The standard format for crontab files in Unix/Linux systems does not use quotes around cron expressions. By supporting this format, the plugin becomes more useful for:

1. Editing actual system crontab files (via `crontab -e`)
2. Working with configuration files that use standard crontab syntax
3. Improving copy-paste compatibility with documentation and examples

## Implementation Details
- Added a new function `cron_from_line_crontab` that:
  - First checks for the existing quoted format
  - Then looks for standard 5-part cron expressions (minute hour day month weekday)
  - Also supports special time strings like @daily, @hourly, etc.
  - Skips comments and empty lines

- The implementation maintains backward compatibility while extending functionality
- Comprehensive test suite ensures both formats work correctly

## Example Configuration
```lua
return {
  'fabridamicelli/cronex.nvim',
  config = function()
    require('cronex').setup({
      -- Add *.crontab to supported file patterns
      file_patterns = { '*.yaml', '*.yml', '*.tf', '*.cfg', '*.config', '*.conf', '*.crontab' },
      
      -- Use custom extractor to support unquoted cron expressions
      extractor = {
        -- Direct reference to the function, not wrapped in another function
        cron_from_line = require("cronex.cron_from_line").cron_from_line_crontab,
        extract = require("cronex.extract").extract
      }
    })
  end
}
```
Screenshot 2025-05-04 at 1 32 13 AM

## Test Plan
1. Run the new acceptance tests:
```
make test
```

Or specifically:
```
nvim --headless --noplugin -u scripts/tests/minimal.vim -c "PlenaryBustedFile tests/acceptance_crontab_spec.lua"
```

2. Manually test with a crontab file containing both quoted and unquoted expressions

### CODE REVIEW

File: tests/acceptance_crontab_spec.lua [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+            
+            -- Wait for async operations to complete
+            wait()
+            
+            -- Get plugin namespace
+            local ns = vim.api.nvim_get_namespaces()["plugin-cronex.nvim"]
+            assert.is_not_nil(ns, "Plugin namespace should exist")
+            
+            -- Get diagnostics
+            local diags = vim.diagnostic.get(buf)
```
Comment (fabridamicelli) [inline tests/acceptance_crontab_spec.lua]: Can we be more specific here with 
```lua
vim.diagnostic.get(buf, { namespace = ns })
```
So that we make sure we look at the diagnostics that we care about. Otherwise the assertion of lines 65,66 could pass with any other present diagnostics.
See [this](https://github.com/fabridamicelli/cronex.nvim/blob/d7c88935e196a0e4dc7b45df898a2aebc115e3cc/tests/acceptance_set_explanations_spec.lua#L34) for a similar example

File: tests/acceptance_crontab_spec.lua [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            assert.is_not_nil(diags, "Diagnostics should exist")
```
Comment (fabridamicelli) [inline tests/acceptance_crontab_spec.lua]: since we have one cron expression (stated in `before_each`) this two tests could be collapsed into one?
```lua
assert.is_true(#diags == 1, "should have 1 diagnostic for cron line")
```

File: tests/acceptance_crontab_spec.lua [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            assert.is_true(#diags > 0, "Should have diagnostics for cron line")
+            
+            -- Verify the diagnostic is on the correct line and has expected content
+            local cron_diag = diags[1]
+            assert.are.equal(0, cron_diag.lnum, "Diagnostic should be on the first line")
+            assert.truthy(string.match(cron_diag.message, "test%-explanation"), 
+                        "Diagnostic message should contain explainer output")
+            assert.are.equal(ns, cron_diag.namespace, "Diagnostic should be in the plugin namespace")
```
Comment (fabridamicelli) [inline tests/acceptance_crontab_spec.lua]: if you add my suggestion for line 63, we can delete this redundant assertion (line 73)

File: tests/acceptance_crontab_spec.lua [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            vim.api.nvim_set_current_buf(buf)
+            local before_diags = vim.diagnostic.get(buf)
+            assert.are.same({}, before_diags, "Should have no diagnostics before enabling")
+            
+            -- Manually enable
+            vim.cmd("CronExplainedEnable")
+            wait()
+            
+            -- Check diagnostics after manual enable
+            local after_diags = vim.diagnostic.get(buf)
```
Comment (fabridamicelli) [inline tests/acceptance_crontab_spec.lua]: same here, it'd be great to check for the cronex-specific diags (by passing namespace opt arg to vim.diagnostic.get

File: tests/acceptance_crontab_spec.lua [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            -- Verify that 6-part cron is recognized correctly
+            for _, diag in ipairs(diags) do
+                if diag.lnum == 7 then
+                    local expected_cron = "1 2 3 4 5" -- 5 parts, ignoring seconds
+                    assert.truthy(string.match(diag.message, "test%-explanation"))
+                end
+            end
+        end)
+        
+        it("Correctly parses 6-part cron expressions", function()
```
Comment (fabridamicelli) [inline tests/acceptance_crontab_spec.lua]: this last `it` is actually more of a unit test (not acceptance), since we are checking the functionality of `cron_from_line_crontab`.
It would be wonderful if we could have it in a separate file `cron_from_file_crontab_spec.lua` (analogous to [this](https://github.com/fabridamicelli/cronex.nvim/blob/main/tests/cron_from_line_spec.lua))

Also, it'd be great to have more extensive examples  (similar to  [this](https://github.com/fabridamicelli/cronex.nvim/blob/main/tests/cron_from_line_spec.lua) to make sure that the regex patterns are working fine on edge cases as well

**Review by fabridamicelli — CHANGES_REQUESTED** [review]
@kylesnowschwartz Thank you very much for this great PR!
I love the detailed explanations and tests.
I'll be happy to merge this as soon as possible since it adds a nice missing functionality.

I left a couple of comments for some fixes. Don't hesitate to ask if you have any questions on my points.
Also, it'd be great if you could sync this branch with the last changes in main (which included formatting instructions with stylua.toml) so that we have everything tidy to release soon a new version:)

File: lua/cronex/cron_from_line.lua [hunk 6]
```diff
@@ -67,4 +67,49 @@ M.cron_from_line = function(line)
     return nil
 end
 
+-- Patterns for standard crontab format (no quotes)
+local crontab_patterns = {
+    -- Standard 5-part cron expression (minute hour day month weekday)
+    "^%s*([%d%*%-%/,]+)%s+([%d%*%-%/,]+)%s+([%d%*%-%/,]+)%s+([%d%*%-%/,]+)%s+([%d%*%-%/,]+)%s",
+    -- Special time strings like @daily, @hourly, etc.
+    "^%s*(@%w+)%s",
```
Comment (kylesnowschwartz) [inline lua/cronex/cron_from_line.lua:75]: Sadly construe doesn't support this format. Do you know of a good cron explainer that does @fabridamicelli ? There are a few options in the Readme but none of them seem like a perfect fit.

**Comment by kylesnowschwartz** [comment]
Thanks @fabridamicelli I've addressed your feedback, please have another read. 

~I have a question as well about the `@` special syntax crontabs support, it seems construe doesn't support that syntax. I wonder if there's a better option for explainers?~

Hmmm construe does support it actually, but I'm getting errors from my testing. I'll have to look deeper into why.

**Review by fabridamicelli — CHANGES_REQUESTED** [review]
@kylesnowschwartz Thank you so much for your changes, that's looks much better.

I just gave it a try locally and found that the extractor is missing the case with days in letters, such as `20 2 * * Sat`. Here's a screenshot using your branch. As you can see on the right, those are indeed valid expressions.


image

It seems like the regexes are not covering that. It'd be great if you could add that and extend the test coverage.

Here the examples in case you want to copy/paste them:
```crontab
30 2 * * Mon command 
30 2 * * Tue command
30 2 * * Wed command
30 2 * * Thu command
30 2 * * Fri command
30 2 * * Sat command
30 2 * * Sun command

30 2 * * mon command
30 2 * * tue command
30 2 * * wed command
30 2 * * thu command
30 2 * * fri command
30 2 * * sat command
30 2 * * sun command
```

**Comment by kylesnowschwartz** [comment]
It required quite a refactor of the parser but I think we got there @fabridamicelli have a look

**Comment by fabridamicelli** [comment]
@kylesnowschwartz thank you so much, it works like a charm now!

I have only a little nitpick for the sake of consistency: the function `is_valid_name` takes currently a the result of `ci(name)` as input, which is neither a name nor a full pattern. Also we build the patterns everytime (concatenating `^.. pat ..$`).
What do you think about building everything at the beginning, like this?:
```lua
local function ci(str)
    local result = ""
    for i = 1, #str do
        local c = str:sub(i, i)
        result = result .. "[" .. c:upper() .. c:lower() .. "]"
    end
    return result
end

local function make_patterns(names)
    local patterns = {}
    for _, name in pairs(names) do
        table.insert(patterns, "^" .. ci(name) .. "$")
    end
    return patterns
end

-- Patterns for field validation
local month_names_patterns = make_patterns({
    "JAN",
    "FEB",
    "MAR",
    "APR",
    "MAY",
    "JUN",
    "JUL",
    "AUG",
    "SEP",
    "OCT",
    "NOV",
    "DEC",
})

local weekday_names_patterns = make_patterns({
    "SUN",
    "MON",
    "TUE",
    "WED",
    "THU",
    "FRI",
    "SAT",
})

```
and later on just calling 
```lua
        for _, pattern in ipairs(name_patterns) do
            if part:match(pattern) then
                return true
            end
        end
```
 
Again, this is just a little detail, otherwise I think we are ready to go. Let me know what you think and I will then merge and make a new release.

Thanks again for your valuable contribution!

**Review by fabridamicelli — CHANGES_REQUESTED** [review]
See last comment

**Comment by kylesnowschwartz** [comment]
> See last comment

Fix in 7a86ff18fd127578bb300ec2b73af66dd45d605b

Thanks for the careful reviews

**Comment by fabridamicelli** [comment]
@kylesnowschwartz thank you. The new release is out :partying_face:
